### PR TITLE
Generate proper, consistent nightly versions for pip and conda packages

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR changes conda Python packages and wheels to all generate a consistent version for nightlies. The nightly version is of the form YY.MM.DDaN, where N is the number of commits from the last tag. The version is embedded in both the package metadata and in the `dask_cuda.__version__` attribute. In addition the commit hash itself is embedded into the package as `dask_cuda.__git_commit__`. These changes ensure that

1. wheels are properly considered nightlies and are treated accordingly by pip (e.g. requiring --pre for installation, not conflicting with normal releases, etc)
2. wheels and conda packages are aligned on versions so that they can be easily compared if necessary.

### CODE REVIEW

File: dask_cuda/VERSION [hunk 1]
```diff
@@ -0,0 +1 @@
+../VERSION
```
Comment (vyasr) [inline dask_cuda/VERSION:1]: The Github UI makes this confusing. You don't want a file containing the text `../VERSION`, this file should be a symlink to the top-level VERSION file.

File: ci/build_python_pypi.sh [hunk 2]
```diff
@@ -11,17 +14,20 @@ export GIT_DESCRIBE_NUMBER=$(git rev-list ${GIT_DESCRIBE_TAG}..HEAD --count)
...
+sed -i "/^__git_commit__/ s/= .*/= \"${commit}\"/g" "${package_name}/_version.py"
+
 # Compute/export RAPIDS_DATE_STRING
 source rapids-env-update
 
 # Update pyproject.toml with pre-release build date
-if ! rapids-is-release-build; then
-  sed -i "s/^version = \""${TOML_VERSION}".*\"/version = \""${PACKAGE_VERSION_NUMBER}"\"/g" pyproject.toml
-fi
+# if ! rapids-is-release-build; then
```
Comment (vyasr) [inline ci/build_python_pypi.sh:28]: You can delete this block now.

File: ci/build_python_pypi.sh [hunk 3]
```diff
@@ -30,6 +36,6 @@ python -m build \
   .
 
 # Revert pyproject.toml pre-release build date
-if ! rapids-is-release-build; then
-  sed -i "s/^version = \""${PACKAGE_VERSION_NUMBER}"\"/version = \""${TOML_VERSION}"\"/g" pyproject.toml
-fi
+# if ! rapids-is-release-build; then
```
Comment (vyasr) [inline ci/build_python_pypi.sh:39]: Delete this block.

**Review by vyasr — CHANGES_REQUESTED** [review]
You also need to specify that the VERSION file should be included in the built distributions. Since dask-cuda uses setuptools directly, you can specify that in [`MANIFEST.in`](https://github.com/rapidsai/dask-cuda/blob/branch-23.12/MANIFEST.in).

File: dask_cuda/VERSION (hunk 1 above)
Comment (jakirkham) [reply dask_cuda/VERSION:1]: This is how GH renders symlinks unfortunately, but good to confirm that this is a symlink

Comment (vyasr) [reply dask_cuda/VERSION]: You can tell that it wasn't a symlink before based on the CI failure.

Comment (galipremsagar) [reply dask_cuda/VERSION]: Thanks for the clarification @vyasr & @jakirkham ! I think I fixed it this time.

File: ci/build_python_pypi.sh (hunk 3 above)
Comment (galipremsagar) [reply ci/build_python_pypi.sh]: Done

File: ci/build_python_pypi.sh (hunk 2 above)
Comment (galipremsagar) [reply ci/build_python_pypi.sh]: Done

**Comment by jakirkham** [comment]
Please see this comment: https://github.com/rapidsai/dask-cuda/pull/1267#pullrequestreview-1702486818

**Comment by galipremsagar** [comment]
> Please see this comment: [#1267 (review)](https://github.com/rapidsai/dask-cuda/pull/1267#pullrequestreview-1702486818)

Thanks for pointing me to this John. @vyasr Apologies I missed it the first time.

**Review by vyasr — APPROVED** [review]

**Review by jakirkham — APPROVED** [review]

**Review by raydouglass — APPROVED** [review]

**Pull request merged**